sklearn/metrics/cluster/unsupervised.py

```diff
@@ -210,6 +210,10 @@ def silhouette_samples(X, labels, metric='euclidean', **kwds):
 
     """
     X, labels = check_X_y(X, labels, accept_sparse=['csc', 'csr'])
+    if metric == 'precomputed':
+        diag_indices = np.diag_indices(X.shape[0])
+        X[diag_indices] = 0
```

Comment:
This modifies the user's data. It either needs a copy, or needs to happen somewhere else. I'd suggested doing it in the chunked reduce operation